hibernate-reactive-core/src/main/java/org/hibernate/reactive/persister/collection/impl/ReactiveBasicCollectionPersister.java

```diff
@@ -260,44 +211,96 @@ protected String getSQLDeleteString() {
 		}
 
 		collection.preInsert( this );
-		// FIXME @Barry: Should we keep it? What's Expectation for?
-		// Expectation expectation = Expectations.appropriateExpectation( getInsertCheckStyle() );
 
 		CompletionStage<Integer> loop = CompletionStages.zeroFuture();
 		ReactiveConnection reactiveConnection = getReactiveConnection( session );
-		Iterator entries = collection.entries( this );
 
-		int i = 0;
+		return CompletionStages.total(
+				entryArray(collection),
+				entry -> loop.thenCompose( total -> reactiveConnection
+						.update( getSQLInsertRowString(), paramValues( entry, collection, id ,session ) )
+						.thenApply( insertCount -> total + insertCount ) ) )
+				.thenAccept( total -> LOG.debugf( "Done inserting rows: %s inserted", total ) );
+	}
+
+
+	/*
+	 * converts the entities of a PersistentCollection into an Object[] array
+	 */
+	private Object[] entryArray(PersistentCollection collection) {
+		List<Object> theList = new ArrayList<>();
+		Iterator entries = collection.entries( this );
 		while ( entries.hasNext() ) {
-			int offset = 1;
 			final Object entry = entries.next();
-			if ( collection.entryExists( entry, i )
-					&& collection.needsInserting( entry, i, elementType ) ) {
+			if ( collection.entryExists( entry, 0 ) ) {
+				theList.add(entry);
+			}
+		}
 
-				// TODO: Check batching: See AbstractCollectionPersister#insertRows
-				// boolean useBatch = expectation.canBeBatched();
+		return theList.toArray( new Object[0] );
+	}
+
+	/*
+	 * converts the entities of a PersistentCollection into an Object[] array for the delete row(s) use case.
+	 */
+	private Object[] entryArray(PersistentCollection collection, boolean deleteByIndex) {
+		List<Object> theList = new ArrayList<>();
+		Iterator deletes = collection.getDeletes( this, !deleteByIndex );
+		while ( deletes.hasNext() ) {
+			final Object entry = deletes.next();
+			if ( collection.entryExists( entry, 0 ) ) {
+				theList.add(entry);
+			}
+		}
 
-				Object[] paramValues = PreparedStatementAdaptor.bind(
-						st -> {
-							int locOffset = writeKey( st, id, offset, session );
+		return theList.toArray( new Object[0] );
+	}
 
-							if ( hasIdentifier ) {
-								locOffset = writeIdentifier( st, collection.getIdentifier( entry, i ), locOffset, session );
-							}
-							if ( hasIndex /* && !indexIsFormula */) {
-								locOffset = writeIndex( st, collection.getIndex( entry, i, this ), locOffset, session );
-							}
-							locOffset = writeElement( st, collection.getElement( entry ), locOffset, session );
+	/*
+	 * generates and returns parameter values for a reactiveRecreate() and reactiveInsertRows() methods
+	 * compatible with getSQLInsertRowString() and query string
+	 */
+	private Object[] paramValues(Object entry, PersistentCollection collection, Serializable id, SharedSessionContractImplementor session) {
+		return PreparedStatementAdaptor.bind(
+				st -> {
+					int i = 0;
+					int loc = writeKey( st, id, 1, session );
```

Comment:
In ORM, offset is initialized like this:
```
offset += expectation.prepare( st );
```
Shouldn't we do the same or something similar?